lib/integrations/awsoidc/externalcloudaudit_iam_config.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awsoidc
+
+import (
+	"context"
+	"net/url"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"github.com/gravitational/trace"
+
+	awslib "github.com/gravitational/teleport/lib/cloud/aws"
+	"github.com/gravitational/teleport/lib/config"
+)
+
+// ConfigureExternalCloudAuditClient is an interface for the AWS client methods
+// used by ConfigureExternalCloudAudit.
+type ConfigureExternalCloudAuditClient interface {
+	PutRolePolicy(context.Context, *iam.PutRolePolicyInput, ...func(*iam.Options)) (*iam.PutRolePolicyOutput, error)
+	GetCallerIdentity(context.Context, *sts.GetCallerIdentityInput, ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error)
+}
+
+// DefaultConfigureExternalCloudAuditClient wraps an iam and sts client to
+// implement ConfigureExternalCloudAuditClient.
+type DefaultConfigureExternalCloudAuditClient struct {
+	Iam *iam.Client
+	Sts *sts.Client
+}
+
+// PutRolePolicy adds or updates an inline policy document that is embedded in
+// the specified IAM role.
+func (d *DefaultConfigureExternalCloudAuditClient) PutRolePolicy(ctx context.Context, input *iam.PutRolePolicyInput, opts ...func(*iam.Options)) (*iam.PutRolePolicyOutput, error) {
+	return d.Iam.PutRolePolicy(ctx, input, opts...)
+}
+
+// GetCallerIdentity returns details about the IAM user or role whose
+// credentials are used to call the operation.
+func (d *DefaultConfigureExternalCloudAuditClient) GetCallerIdentity(ctx context.Context, input *sts.GetCallerIdentityInput, opts ...func(*sts.Options)) (*sts.GetCallerIdentityOutput, error) {
+	return d.Sts.GetCallerIdentity(ctx, input, opts...)
+}
+
+// ConfigureExternalCloudAudit attaches an IAM policy with necessary permissions
+// for the ExternalCloudAudit feature to an existing IAM role associated with an
+// AWS OIDC integration.
+func ConfigureExternalCloudAudit(
```

Comment:
:nail_care: 
It seems that we are not following the `CheckAndSet and then Do The Thing` pattern.
So, the tests look a bit weird because we need to mutate the internal state of the request (which is usually done in the CheckAndSetDefaults).

No big deal I think, because this method should not evolve too much.

Reply:
I don't quite follow, where is the internal state of the request being mutated? I could add a CheckAndSetDefaults but it seemed redundant because the CLI lib already enforces that all fields are required

Reply:
Sorry, I was mixing the `policyCfg` with the `params`
I still think that it deserves a `CheckAndSetDefaults` because we are parsing the URIs, but up to you :+1: